apis/supernova/src/hooks.js

```diff
@@ -979,3 +985,111 @@ export function onTakeSnapshot(cb) {
   }
   h.fn = cb;
 }
+
+/**
+ * @experimental
+ * @interface Keyboard
+ * @property {boolean} enabled Whether or not Nebula handles keyboard navigation or not.
+ * @property {boolean} active Set to true when the chart is activated, ie a user tabs to the chart and presses Enter or Space.
+ * @property {function=} blur Function used by the visualization to tell Nebula to it wants to relinquish focus
+ * @property {function=} focus Function used by the visualization to tell Nebula to it wants focus
+ */
+
+/**
+ * Gets the desired keyboard settings and status to applied when rendering the visualization.
+ * A visualization should in general only have tab stops if either `keyboard.enabled` is false or if active is true.
+ * This means that either Nebula isn't configured to handle keyboard input or the chart is currently focused.
+ * Enabling or disabling keyboardNavigation are set on the embed configuration and
+ * should be respected by the visualization.
+ * @entry
+ * @returns {Keyboard}
+ * @example
+ * // configure nebula to enable navigation between charts
+ * embed(app, {
+ *   context: {
+ *     keyboardNavigation: true, // tell Nebula to handle navigation
+ *   }
+ * }).render({ element, id: 'sdfsdf' });
+ *
+ * @example
+ * import { useKeyboard } from '@nebula.js/stardust';
+ * // ...
+ * const keyboard = useKeyboard();
+ * useEffect(() => {
+ *  // Set a tab stop on our button if in focus or if Nebulas navigation is disabled
+ *  button.setAttribute('tabIndex', keyboard.active || !keyboard.enabled ? 0 : -1);
+ *  // If navigation is enabled and focus has shifted, lets focus the button
+ *  keyboard.enabled && keyboard.active && button.focus();
+ * }, [keyboard])
+ *
+ */
+
+export function useKeyboard() {
+  const keyboardNavigation = useInternalContext('keyboardNavigation');
+  const blurCallback = useInternalContext('blurCallback');
+
+  if (!currentComponent.__hooks.accessibility.exitFunction) {
+    const exitFunction = function (resetFocus) {
+      const acc = this.__hooks.accessibility;
+      if (acc.enabled && acc.active) {
+        blur(this);
+        blurCallback && blurCallback(resetFocus);
+      }
+    }.bind(currentComponent);
+
+    currentComponent.__hooks.accessibility.exitFunction = exitFunction;
+
+    const focusFunction = function () {
+      const acc = this.__hooks.accessibility;
+      if (acc.enabled && !acc.active) {
+        blurCallback && blurCallback(false);
+        focus(this);
+      }
+    }.bind(currentComponent);
+
+    currentComponent.__hooks.accessibility.focusFunction = focusFunction;
+  }
```

Comment:
I guess you need to check if focusFunction is set as well?

Reply:
not really, they are both created inside that IF, so if one is missing then both are and vice-versa

Reply:
ok good :)